### PR TITLE
ASA examples

### PR DESCRIPTION
**Pull request by Invertisment**
Depends on #65

### CODE REVIEW

File: examples/deployment/artifacts/scripts/0-gold-asa.js.cp.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+default:
```
Comment (Invertisment) [inline examples/deployment/artifacts/scripts/0-gold-asa.js.cp.yaml:1]: Will remove these files

File: examples/deployment/algob.config.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+//console.log("hi", ab);
+//const { mkAccounts } = require("algorand-builder");
+
+const { mkAccounts } = require("algob");
+let accounts = mkAccounts([{
+  name: "master-account",
+  // goal account list -d ~/.algorand-local/Node/
+  addr: "WWYNX3TKQYVEREVSW6QQP3SXSFOCE3SKUSEIVJ7YAGUPEACNI5UGI4DZCE",
+  // Use your own account.
+  // Export using:
```
Comment (robert-zaremba) [inline examples/deployment/algob.config.js:25]: let's be more verbose here:

```
// Exported from algorand node using an account with ALOGs (WWY...)
// ...
```

File: examples/deployment/.editorconfig [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+root = true
+
+[*]
+charset = utf-8
+end_of_line = lf
+indent_style = space
+indent_size = 2
+insert_final_newline = true
+trim_trailing_whitespace = false
+max_line_length = 96
```
Comment (robert-zaremba) [inline examples/deployment/.editorconfig:11]: :+1:

File: examples/deployment/README.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+ASA deployment using Algorand builder from Scale it
+
+Needs to be linked (`yarn link`) with algob:
+https://github.com/scale-it/algorand-builder/
+
+Authors: Invertisment & amityadav0
```
Comment (robert-zaremba) [inline examples/deployment/README.md:6]: this is a whole team effort, but you can link your repository if you are planning to do more updates.

File: examples/deployment/README.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+ASA deployment using Algorand builder from Scale it
```
Comment (robert-zaremba) [inline examples/deployment/README.md:1]: how about removing a scale it and use a link to the repo in `Algorand Builder`?

File: examples/deployment/algob.config.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+// const { loadAccountsFromFileSync } = require("algob");
+// // Example: accounts loaded from a file:
+// const accFromFile = loadAccountsFromFileSync("assets/accounts_generated.yaml");
+// accounts = accounts.concat(accFromFile);
+
+// Node commands:
+// Start whole network:
+// goal network start -r ~/.algorand-local/
+// Start/stop a single node:
+// goal node start -d ~/.algorand-local/Node/
```
Comment (robert-zaremba) [inline examples/deployment/algob.config.js:55]: Let move this comments to README

File: examples/deployment/artifacts/scripts/0-gold-asa.js.cp.yaml (hunk 1 above)
Comment (robert-zaremba) [reply examples/deployment/artifacts/scripts/0-gold-asa.js.cp.yaml]: yes, thanks

File: examples/deployment/scripts/query/john-balances.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  const teslaAssetID = deployer.asa.get("tesla").assetIndex
+
+  const johnAccount = deployer.accountsByName.get("john-account");
+
+  // print one by one
+  //await printAssetHolding(deployer, johnAccount.addr, goldAssetID);
+  //await printAssetHolding(deployer, johnAccount.addr, teslaAssetID);
+  // print all at once
+  await printAssets(deployer, johnAccount.addr)
+}
```
Comment (robert-zaremba) [inline examples/deployment/scripts/query/john-balances.js:22]: :+1:

File: examples/deployment/scripts/transfer/gold-to-john.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+const algosdk = require('algosdk');
+const {
+  printCreatedAsset,
+  printAssetHolding,
+  transferMicroAlgos,
+  asaOptIn,
+  transferAsset
+} = require('../../src/asa-helpers');
+
+async function run(runtimeEnv, accounts, deployer) {
```
Comment (robert-zaremba) [inline examples/deployment/scripts/transfer/gold-to-john.js:10]: let's add a comment above the script, explaining that each execution through `algob run` will create and execute transfer transaction....

File: examples/deployment/scripts/transfer/master-fund-john.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+const algosdk = require('algosdk');
+const {
+  printCreatedAsset,
+  printAssetHolding,
+  transferMicroAlgos,
+  asaOptIn,
+  transferAsset
+} = require('../../src/asa-helpers');
+
+async function run(runtimeEnv, accounts, deployer) {
```
Comment (robert-zaremba) [inline examples/deployment/scripts/transfer/master-fund-john.js:10]: same here, let's add some docs comments.

File: examples/deployment/src/asa-helpers.js [hunk 10]
```diff
@@ -0,0 +1,163 @@
+// Authors of example: Invertisment & amityadav0
+const algosdk = require('algosdk');
+
+// Function used to print created asset for account and assetid
+exports.printCreatedAsset = async function (deployer, account, assetid) {
```
Comment (robert-zaremba) [inline examples/deployment/src/asa-helpers.js:5]: the function doesn't need deployer, it requires only `algodClient`. 
Let's update the parameter.

File: examples/deployment/src/asa-helpers.js [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  // note: if you have an indexer instance available it is easier to just use this
+  //     let accountInfo = await indexerClient.searchAccounts()
+  //    .assetID(assetIndex).do();
+  // and in the loop below use this to extract the asset for a particular account
+  // accountInfo['accounts'][idx][account]);
+  let accountInfo = await deployer.algodClient.accountInformation(account).do();
+  for (idx = 0; idx < accountInfo['created-assets'].length; idx++) {
+    let scrutinizedAsset = accountInfo['created-assets'][idx];
```
Comment (robert-zaremba) [inline examples/deployment/src/asa-helpers.js:13]: let's use `for (asset in accontInfo['created-assets'])` - this case you don't need to play with indexes.

Side note: you didn't want to use `forEach`? ;)

File: examples/deployment/src/asa-helpers.js [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    let scrutinizedAsset = accountInfo['assets'][idx];
+    if (scrutinizedAsset['asset-id'] == assetid) {
+      console.log("Asset Holding Info:", scrutinizedAsset, accountInfo);
+      break;
+    }
+  }
+};
+
+// Print account's ASA info. All at once.
+exports.printAssets = async function (deployer, account) {
```
Comment (robert-zaremba) [inline examples/deployment/src/asa-helpers.js:42]: as above, let's use  `algodClient` directly. same for other functions below.

File: examples/deployment/src/asa-helpers.js [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+    amount: amount,
+    assetID: assetID,
+    txId: xtx.txId
+  })
+  // wait for transaction to be confirmed
+  await deployer.waitForConfirmation(xtx.txId);
+
+  //console.log("Master Account = " + master);
+  //await printAssetHolding(deployer.algodClient, master, assetID);
+}
```
Comment (robert-zaremba) [inline examples/deployment/src/asa-helpers.js:163]: how about adding most of this functions to `algob` package? Eg - they can be in `src/lib/printers.ts`

The transfer functions also should be in `algob` but in other module.

File: packages/algob/src/lib/script-checkpoints.ts [hunk 14]
```diff
@@ -176,18 +176,27 @@ export function persistCheckpoint (scriptName: string, checkpoint: Checkpoints):
...
-      cp[cpKey] = cpVal;
-    }
-    obj[key] = cp;
+  const checkpoints = loadFromYamlFileSilent(checkpointName, { mapAsMap: false });
+  for (const k of Object.keys(checkpoints)) {
+    convertCPValsToMaps(checkpoints[k]);
   }
-  return obj;
+  return checkpoints;
 }
```
Comment (robert-zaremba) [inline packages/algob/src/lib/script-checkpoints.ts:200]: :+1:

**Review by robert-zaremba — APPROVED** [review]
Small changes required.

File: examples/deployment/README.md (hunk 4 above)
Comment (Invertisment) [reply examples/deployment/README.md:6]: I think it's good to have it in a repository but I'm not sure how to handle the updates then. 
I can mention it. 
I'll change this line.

File: examples/deployment/src/asa-helpers.js (hunk 11 above)
Comment (Invertisment) [reply examples/deployment/src/asa-helpers.js:13]: This is mostly Amit's code so I added the pieces together without changing too much

Comment (robert-zaremba) [reply examples/deployment/src/asa-helpers.js:13]: His code was for training, it was good, but now in the repo we can clean few bits.

File: examples/deployment/README.md (hunk 4 above)
Comment (Invertisment) [reply examples/deployment/README.md:6]: I'll add all 4 people.

File: examples/deployment/README.md (hunk 5 above)
Comment (Invertisment) [reply examples/deployment/README.md:1]: :+1:

File: examples/deployment/algob.config.js (hunk 2 above)
Comment (Invertisment) [reply examples/deployment/algob.config.js:25]: :+1:

File: examples/deployment/algob.config.js (hunk 6 above)
Comment (Invertisment) [reply examples/deployment/algob.config.js:55]: :+1:

File: examples/deployment/scripts/transfer/gold-to-john.js (hunk 8 above)
Comment (Invertisment) [reply examples/deployment/scripts/transfer/gold-to-john.js:10]: I'll add this:
```
// Create "gold" Algorand Standard Asset (ASA)
// Accounts are loaded from config
// To use ASA accounts have to opt-in and owner is opt-in by default
// john-account is transferred some funds to execute opt-in transaction
```

File: examples/deployment/README.md (hunk 5 above)
Comment (Invertisment) [reply examples/deployment/README.md:1]: Changed

File: examples/deployment/algob.config.js (hunk 6 above)
Comment (Invertisment) [reply examples/deployment/algob.config.js:55]: moved

File: examples/deployment/src/asa-helpers.js (hunk 10 above)
Comment (Invertisment) [reply examples/deployment/src/asa-helpers.js:5]: I can't use algodClient directly. It doesn't have `waitForConfirmation` function.

File: examples/deployment/scripts/transfer/master-fund-john.js (hunk 9 above)
Comment (Invertisment) [reply examples/deployment/scripts/transfer/master-fund-john.js:10]: added same as in previous one

File: examples/deployment/src/asa-helpers.js (hunk 12 above)
Comment (Invertisment) [reply examples/deployment/src/asa-helpers.js:31]: In some functions I used `Deployer.waitForConfirmation` so it's better not to mix

File: examples/deployment/src/asa-helpers.js (hunk 13 above)
Comment (Invertisment) [reply examples/deployment/src/asa-helpers.js:140]: What about later?

File: examples/deployment/src/asa-helpers.js (hunk 11 above)
Comment (Invertisment) [reply examples/deployment/src/asa-helpers.js:7]: Changed

File: examples/deployment/src/asa-helpers.js (hunk 12 above)
Comment (robert-zaremba) [reply examples/deployment/src/asa-helpers.js:30]: OK

File: examples/deployment/README.md [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+### Run
+`algob deploy`
+`algob run scripts/query/john-balances.js`
+
+## Authors
+- robert-zaremba
+- Invertisment
+- amityadav0
+- as04
```
Comment (robert-zaremba) [inline examples/deployment/README.md:30]: Do we need this at all? If you want it then you should put yourself first, or use alphabetic order and remove `as04` - there is 0 contribution / working code from her.

Comment (Invertisment) [reply examples/deployment/README.md:30]: I think I should put something in my own repository. 
Will remove from both places then.

**Pull request merged**